docs/developer/design-v2-2023-11-11.md

```diff
@@ -0,0 +1,198 @@
+-   To enable a high degree of automation in resource provisioning and management, thereby reducing manual overhead and potential for errors.
+
+### Features
+
+-   **Single Authentication**: Utilizing an app password to access and manage resources without the need for multiple credentials.
+-   **Global and Regional Resource Management**: Allowing users to manage resources that are not bound by regional constraints, as well as those that are region-specific.
+-   **Dynamic Client Allocation**: Depending on the resource's requirement, the appropriate client (DB, Frontegg, Cloud API and etc.) will be allocated dynamically.
+
+## Architecture
+
+### Provider Configuration
+
+The provider will be initialized with parameters that determine its behavior and default settings:
+
+```hcl
+provider "materialize" {
+  app_password   = var.materialize_app_password
+  default_region = var.materialize_default_region
+  endpoint       = var.materialize_endpoint
+}
+```
+
+Users will specify the app password and the default region for operations. Optionally, the endpoint can be overridden for testing or staging purposes.
+
+### Multi-Client Structure
+
+The provider will maintain a map of clients for database operations across different regions and a client for interfacing with the Frontegg API.
+
+```hcl
+type ProviderMeta struct {
+    DB       map[Region]*clients.DBClient
+    Frontegg *clients.FronteggClient
+    // Future: CloudAPI *clients.CloudAPIClient
+}
+```
+
+This structure allows for scalability and extension of the provider to incorporate additional clients as necessary.
+
+## Resources and Operations
+
+### Region Resource
+
+Responsible for managing the availability of Materialize regions. It will interact with the Cloud API to enable or retrieve regions as requested by the user.
+
+```hcl
+resource "materialize_region" "aws_us_east_1" {
+  region = "aws/us-east-1"
+}
+```
+
+Operations:
+
+-   **Enable Region**: Uses the Cloud API to enable a new region.
+-   **Read Region**: Retrieves the status of a region from the Cloud API.
+-   **Disable Region**: Noop. Regions cannot be disabled once enabled.
+
+### User Resource
+
+Manages the lifecycle of a user within the Frontegg platform.
+
+```hcl
+resource "materialize_app_user" "example_user" {
+  email          = "user@example.com"
+  auth_provider  = "local"
+  role_ids       = ["role-id-1", "role-id-2"] # Logic to retrieve role IDs from role names?
+  # Additional attributes...
+}
+```
+
+Operations:
+
+-   **Create User**: Invokes the Frontegg API to create/invite a user.
+-   **Read User**: Retrieves user details from Frontegg.
+-   **Update User**: Attaches or detaches roles from the user.
+-   **Delete User**: Removes the user from Frontegg.
+
+### Organization Data Source?
+
+Generates a list of available organizations from Frontegg.
+
+```hcl
+data "materialize_app_organizations" "all" {}
```

Comment:
App passwords are specific to a single organization, so if we want a data source here, it should just be for the single organization that the app password is tied to.